File: java/TypeDB.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.typedb.driver;

import com.typedb.driver.api.Credentials;
import com.typedb.driver.api.Driver;
import com.typedb.driver.api.DriverOptions;
import com.typedb.driver.api.concept.GivenRows;
import com.typedb.driver.api.concept.value.Value;
import com.typedb.driver.common.Duration;
import com.typedb.driver.common.exception.TypeDBDriverException;
import com.typedb.driver.concept.GivenRowsImpl;
import com.typedb.driver.concept.value.ValueImpl;
import com.typedb.driver.connection.DriverImpl;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class TypeDB {
    public static final String DEFAULT_ADDRESS = "127.0.0.1:1729";

    /**
     * Open a TypeDB Driver to a TypeDB server available at the provided address.
     *
     * <h3>Examples</h3>
     * <pre>
     * TypeDB.driver(address);
     * </pre>
     *
     * @param address       The address of the TypeDB server
     * @param credentials   The credentials to connect with
     * @param driverOptions The driver connection options to connect with
     */
    public static Driver driver(String address, Credentials credentials, DriverOptions driverOptions) throws TypeDBDriverException {
        return new DriverImpl(address, credentials, driverOptions);
    }

    /**
     * Open a TypeDB Driver to a TypeDB cluster available at the provided addresses.
     *
     * <h3>Examples</h3>
     * <pre>
     * TypeDB.driver(address);
     * </pre>
     *
     * @param addresses     The addresses of TypeDB cluster servers for connection
     * @param credentials   The credentials to connect with
     * @param driverOptions The driver connection options to connect with
     */
    public static Driver driver(Set<String> addresses, Credentials credentials, DriverOptions driverOptions) throws TypeDBDriverException {
        return new DriverImpl(addresses, credentials, driverOptions);
    }

    /**
     * Open a TypeDB Driver to a TypeDB cluster, using the provided address translation.
     *
     * <h3>Examples</h3>
     * <pre>
     * TypeDB.driver(addresses);
     * </pre>
     *
     * @param addressTranslation The translation of public TypeDB cluster server addresses (keys) to server-side private addresses (values)
     * @param credentials        The credentials to connect with
     * @param driverOptions      The driver connection options to connect with
     */
    public static Driver driver(Map<String, String> addressTranslation, Credentials credentials, DriverOptions driverOptions) throws TypeDBDriverException {
        return new DriverImpl(addressTranslation, credentials, driverOptions);
    }

    /**
     * Factory class to instantiate new <code>Value</code> concepts from raw values.
     * These can then be used in the <code>GivenRows</code> of a query.
     */
    public static class Concept {

        /**
         * Constructs a <code>GivenRows</code> instance for use as inputs to queries.
         *
         * @param variables The variables describing the content of the givenRows.
         * @param rows Input rows for the query; each inner iterable is one row, {@code Optional.empty()} entries represent empty variables.
         */
        public static GivenRows givenRows(List<String> variables, List<? extends List<? extends com.typedb.driver.api.concept.Concept>> rows) throws TypeDBDriverException {
            return GivenRowsImpl.of(variables, rows);
        }

        /**
         * Constructs a <code>GivenRows</code> instance from the dictionary for use as inputs to queries.
         *
         * @param givenRows A list of input rows for the query. Each row is a dictionary mapping a variable to its value.
         */
        public static GivenRows givenRows(List<? extends Map<String, ? extends com.typedb.driver.api.concept.Concept>> givenRows) throws TypeDBDriverException {
            return GivenRowsImpl.of(givenRows);
        }

        /**
         * Constructs a <code>GivenRows</code> instance from a list of rows given as plain Java objects.
         * Each map entry value may be a {@link com.typedb.driver.api.concept.Concept} (used directly) or a raw host-language value
         * which is converted via {@link #tryConvertToValue}.
         *
         * @param givenRows A list of input rows; each row maps variable names to concepts or raw values.
         */
        public static GivenRows givenRowsFromObjects(List<? extends Map<String, Object>> givenRows) throws TypeDBDriverException {
            return GivenRowsImpl.ofObjects(givenRows);
        }

        /**
         * Converts a raw host-language object to a {@link Value} concept.
         * Accepted types: {@code Boolean}, {@code Long}, {@code Integer}, {@code Double}, {@code Float},
         * {@code BigDecimal}, {@code String}, {@code LocalDate}, {@code LocalDateTime}, {@code ZonedDateTime}, {@code Duration}.
         *
         * @throws TypeDBDriverException if the object's type is not supported
         */
        public static Value tryConvertToValue(Object value) {
            return ValueImpl.tryConvertToValue(value);
        }

        /** Creates a new {@code Value} wrapping the specified {@code boolean} value. */
        public static Value newBoolean(boolean value) {
            return ValueImpl.newBoolean(value);
        }

        /** Creates a new {@code Value} wrapping the specified {@code long} value. */
        public static Value newInteger(long value) { return ValueImpl.newInteger(value); }

        /** Creates a new {@code Value} wrapping the specified {@code double} value. */
        public static Value newDouble(double value) { return ValueImpl.newDouble(value); }

        /** Creates a new {@code Value} wrapping the specified {@code BigDecimal} value. */
        public static Value newDecimal(BigDecimal value) { return ValueImpl.newDecimal(value); }

        /** Creates a new {@code Value} wrapping the specified {@code String} value. */
        public static Value newString(String value) { return ValueImpl.newString(value); }

        /** Creates a new {@code Value} wrapping the specified {@code LocalDate} value. */
        public static Value newDate(LocalDate value) { return ValueImpl.newDate(value); }

        /** Creates a new {@code Value} wrapping the specified {@code LocalDateTime} value. */
        public static Value newDatetime(LocalDateTime value) { return ValueImpl.newDatetime(value); }

        /** Creates a new {@code Value} wrapping the specified {@code ZonedDateTime} value. */
        public static Value newDatetimeTz(ZonedDateTime value) { return ValueImpl.newDatetimeTz(value); }

        /** Creates a new {@code Value} wrapping the specified {@code Duration} value. */
        public static Value newDuration(Duration value) { return ValueImpl.newDuration(value); }
    }
}

File: csharp/TypeDB.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using TypeDB.Driver.Api;
using TypeDB.Driver.Common;
using TypeDB.Driver.Connection;
using ConceptGivenRows = TypeDB.Driver.Concept.GivenRows;
using ConceptValue = TypeDB.Driver.Concept.Value;

namespace TypeDB.Driver
{
    public static class TypeDB
    {
        /// <summary>
        /// The default address of the TypeDB server.
        /// </summary>
        public const string DefaultAddress = "127.0.0.1:1729";

        /// <summary>
        /// Open a TypeDB Driver to a TypeDB server available at the provided address.
        /// </summary>
        /// <param name="address">The address of the TypeDB server.</param>
        /// <param name="credentials">The credentials to connect with.</param>
        /// <param name="driverOptions">The driver connection options to connect with.</param>
        /// <example>
        /// <code>
        /// TypeDB.Driver(address, credentials, driverOptions);
        /// </code>
        /// </example>
        public static IDriver Driver(string address, Credentials credentials, DriverOptions driverOptions)
        {
            return new TypeDBDriver(address, credentials, driverOptions);
        }

        /// <summary>
        /// Open a TypeDB Driver to a TypeDB cluster available at the provided addresses.
        /// </summary>
        /// <param name="addresses">The addresses of TypeDB cluster servers for connection.</param>
        /// <param name="credentials">The credentials to connect with.</param>
        /// <param name="driverOptions">The driver connection options to connect with.</param>
        /// <example>
        /// <code>
        /// TypeDB.Driver(addresses, credentials, driverOptions);
        /// </code>
        /// </example>
        public static IDriver Driver(ISet<string> addresses, Credentials credentials, DriverOptions driverOptions)
        {
            return new TypeDBDriver(addresses, credentials, driverOptions);
        }

        /// <summary>
        /// Open a TypeDB Driver to a TypeDB cluster, using the provided address translation.
        /// </summary>
        /// <param name="addressTranslation">The translation of public TypeDB cluster server addresses (keys) to server-side private addresses (values).</param>
        /// <param name="credentials">The credentials to connect with.</param>
        /// <param name="driverOptions">The driver connection options to connect with.</param>
        /// <example>
        /// <code>
        /// TypeDB.Driver(addressTranslation, credentials, driverOptions);
        /// </code>
        /// </example>
        public static IDriver Driver(IDictionary<string, string> addressTranslation, Credentials credentials, DriverOptions driverOptions)
        {
            return new TypeDBDriver(addressTranslation, credentials, driverOptions);
        }

        /// <summary>
        /// Factory class to instantiate new <c>IValue</c> concepts and build <c>IGivenRows</c> for use as query inputs.
        /// </summary>
        public static class Concept
        {
            /// <summary>Constructs an <c>IGivenRows</c> instance for use as inputs to queries (index-based).</summary>
            public static IGivenRows GivenRows(List<string> givenVariables, List<List<IConcept?>> givenRows)
                => ConceptGivenRows.Of(givenVariables, givenRows);

            /// <summary>Constructs an <c>IGivenRows</c> instance from dict-based rows for use as inputs to queries.</summary>
            public static IGivenRows GivenRows(List<Dictionary<string, IConcept?>> givenRows)
                => ConceptGivenRows.Of(givenRows);

            /// <summary>Constructs an <c>IGivenRows</c> instance from rows containing raw .NET values or concepts.</summary>
            public static IGivenRows GivenRowsFromObjects(List<Dictionary<string, object?>> givenRows)
                => ConceptGivenRows.OfObjects(givenRows);

            /// <summary>
            /// Converts a raw host-language object to an <c>IValue</c> concept.
            /// Accepted types: <c>bool</c>, <c>long</c>, <c>int</c>, <c>double</c>, <c>float</c>,
            /// <c>decimal</c>, <c>string</c>, <c>DateOnly</c>, <c>Datetime</c>, <c>DatetimeTZ</c>, <c>Duration</c>.
            /// </summary>
            /// <exception cref="TypeDBDriverException">if the object's type is not supported.</exception>
            public static IValue TryConvertToValue(object value) => ConceptValue.TryConvertToValue(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>bool</c> value.</summary>
            public static IValue NewBoolean(bool value) => ConceptValue.NewBoolean(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>long</c> value.</summary>
            public static IValue NewInteger(long value) => ConceptValue.NewInteger(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>double</c> value.</summary>
            public static IValue NewDouble(double value) => ConceptValue.NewDouble(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>decimal</c> value.</summary>
            public static IValue NewDecimal(decimal value) => ConceptValue.NewDecimal(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>string</c> value.</summary>
            public static IValue NewString(string value) => ConceptValue.NewString(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>DateOnly</c> value.</summary>
            public static IValue NewDate(DateOnly value) => ConceptValue.NewDate(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>Datetime</c> value.</summary>
            public static IValue NewDatetime(Datetime value) => ConceptValue.NewDatetime(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>DatetimeTZ</c> value.</summary>
            public static IValue NewDatetimeTz(DatetimeTZ value) => ConceptValue.NewDatetimeTz(value);

            /// <summary>Creates a new <c>IValue</c> wrapping the specified <c>Duration</c> value.</summary>
            public static IValue NewDuration(Duration value) => ConceptValue.NewDuration(value);
        }
    }
}
